src/utils/waitFor.ts

```diff
@@ -0,0 +1,46 @@
+import { Atom, Getter } from '../core/types'
+import { atomFamily } from './atomFamily'
+
+export const waitForAll = atomFamily(
+  (dependencies: ReadonlyArray<Atom<unknown>>) => (get) => {
+    const [results, exceptions] = concurrentRequests(get, dependencies)
+
+    if (exceptions.every((exception) => exception == null)) {
+      return results
+    }
+
+    const error = exceptions.find(
+      (exception) => exception != null && !(exception instanceof Promise)
+    )
+
+    if (error != null) {
+      throw error
+    }
+
+    return Promise.all(exceptions).then((exceptionResults) =>
+      exceptionResults.map((r, index) => {
+        r === undefined ? results[index] : r
+      })
+    )
+  }
+)
+
+function concurrentRequests(
+  get: Getter,
+  dependencies: ReadonlyArray<Atom<unknown>>
+) {
+  const results = Array<unknown | undefined>(dependencies.length).fill(
+    undefined
+  )
```

Comment:
```suggestion
  const results = []
```
Shouldn't this be fine?

Reply:
No, because I want to keep the order of the atom results, that's why I fill two arrays and then stitch them together at the end to keep the order.